Self healing proposal (#3)
ShubhamRwt/proposals: 090-self-healing-using-cruise-control.md

# Self Healing using Cruise Control

This proposal is about integrating the Self-healing feature of Cruise Control into the Strimzi operator.
The Self-healing feature of Cruise Control allows us to heal the anomalies detected in the cluster automatically. 

> Anomaly are detected by Cruise Control using the Anomaly detector classes 

Reviewer: "anomaly detector manager" I would say. Remember, no details or language referring to classes, interfaces or whatever when you are explaining an high level picture of something.

## Current situation

> During normal operations it common for Kafka clusters to become unbalanced (through partition key skew, uneven partition placement etc.) or for problems to occur with disks or other underlying hardware which makes the cluster unhealthy.  

Reviewer: ```suggestion
During normal operations it's common for Kafka clusters to become unbalanced (through partition key skew, uneven partition placement etc.) or for problems to occur with disks or other underlying hardware which makes the cluster unhealthy.  
```

> Currently, if we encounter any such scenario we need to fix these issues manually i.e. if there is some broker failure then we might move the partition replicas form that corrupted broker to some other healthy broker by using the `KafkaRebalance` custom resource in the `remove-broker` mode.

Reviewer: ```suggestion
Currently, if we encounter any such scenario we need to fix these issues manually i.e. if there is some broker failure then we might move the partition replicas from that corrupted broker to some other healthy broker by using the `KafkaRebalance` custom resource in the `remove-broker` mode.
```

While with small sized cluster it is feasible to fix things manually, when it comes to larger ones it can be very time-consuming or just not feasible to fix all the issue on your own.
Also, it is possible that while doing thing manually, that we might lose data or cause other unforeseen issues with the Kafka Cluster.

## Motivation

With the help of the self-healing feature we can resolve the issues like disk failure, broker failure and other issues automatically in the cluster.
Cruise Control treats these issues as "anomalies", the detection of which is the responsibility of the anomaly detector mechanism.

> Currently, Strimzi has the support to enable anomaly detection by using a custom notifier but the missing part is the interaction between the operator and Cruise Control. Also, even if the anomaly will be detected, the rebalancing would be done manually by using the `KafkaRebalance` custom resource

Reviewer: "the missing part is the interaction between the operator and Cruise Control" ... it's not valid anymore, we don't need such interaction.

> It would be great to have a way to keep fixing the anomalies whenever they are detected automatically

Reviewer: ```suggestion
It would be great to have a way to keep fixing the anomalies whenever they are detected automatically.
```

Reviewer 2: ```suggestion
It would be useful for user of Strimzi to be able to have these anomalies fixed automatically whenever they are detected.
```

## Proposal

> This proposal is about introducing the self-healing feature of Cruise Control into the strimzi operator.

Reviewer 2: ```suggestion
This proposal is about introducing the self-healing feature of Cruise Control into the Strimzi operator.
```

This feature would allow the cluster to be healed automatically if there are some anomalies detected.

> ### Introduction

Reviewer 2: A meta-comment about the introduction. I think is a good that your are describing how self-healing works, but I still thinks some flow-charts/diagrams would aid understanding of the operation-flow. The easiest way would probably be to create [mermaid diagrams](https://github.blog/developer-skills/github/include-diagrams-markdown-files-mermaid/) directly in the markdown file.

#### Anomaly Detection

> Anomaly detection in Cruise Control makes use various anomaly detection classes to detect the anomaly.

Reviewer 2: ```suggestion
Anomaly detection in Cruise Control makes use of various detection classes.
```

Reviewer 2: What do these do? Do they have trigger logic, do they use signals from sensors? How do they detect anomalies?

> The `AnomalyDetectorManager` class is then prompted regarding the detected anomalies and is responsible to handle these detected anomalied i.e whether to fix the anomaly or not.

Reviewer 2: ```suggestion
Once a detection class is triggered, the `AnomalyDetectorManager` class is then prompted with the details of the anomalies and is responsible for handling the healing of them i.e whether to fix the anomaly or not.
```
Is is plural "anomalies" or does the manager only get passed one "anomaly" at a time?

> The anomalies detected can be of various types: broker-failure, disk-failure, topic anomalies, metric anomalies and maintenance anomalies.

Reviewer 2: ```suggestion
The detected anomalies can be of various types: broker-failure, disk-failure, topic anomalies, metric anomalies and maintenance anomalies.
```

Reviewer 2: You should probably cover this at the begining, what are the anomaly types, how do the detectors detect them etc?

> There are also multiple anomaly detection related [configurations](https://github.com/linkedin/cruise-control/wiki/Configurations#anomalydetector-configurations) like the anomaly detection interval for anomalies, the anomaly notifier class etc. which can be configured to make anomaly detection more efficient

Reviewer 2: ```suggestion
Anomaly detection also has several [configurations](https://github.com/linkedin/cruise-control/wiki/Configurations#anomalydetector-configurations), such as the detection interval and the anomaly notifier class, which can effect the performance of the Cruise Control server and the latency of the anomaly detection.
```

#### Notifiers in Cruise Control

> Whenever anomalies are detected, then Cruise Control provides ability to notify the user regarding the detected anomalies using the notifier classes.

Reviewer 2: ```suggestion
Whenever anomalies are detected, Cruise Control provides the ability to notify the user regarding the detected anomalies using optional notifier classes.
```

> The notification sent by these notifiers increases the visibility of the operation that are taken by Cruise Control.

Reviewer 2: ```suggestion
The notification sent by these classes increases the visibility of the operation that are taken by Cruise Control.
```

> Cruise Control by default uses `NoopNotifier` but this notifier is configurable and custom user-built notifiers can be used by using the `anomaly.notifier.class` property

Reviewer 2: ```suggestion
By default, Cruise Control uses the `NoopNotifier` which silently declines to action any detected anomaly. 
However, the notifier class used by Cruise Control is configurable and custom notifiers can be used by setting the `anomaly.notifier.class` property.
```

Reviewer 2: Does the NoopNotifier silently approve if self-healing is enabled and silently decline if it is disabled?

Reviewer: No, the NoopNotifier just returns IGNORE in any case.

> Cruise Control also provides some [custom notifiers](https://github.com/linkedin/cruise-control/wiki/Configure-notifications) like Slack Notifier, Alerta Notifier etc. There are multiple other[Self-healing notifier](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations) related configurations you can use to make notfier more efficient as per the use case

Reviewer: ```suggestion
Cruise Control also provides some [custom notifiers](https://github.com/linkedin/cruise-control/wiki/Configure-notifications) like Slack Notifier, Alerta Notifier etc. There are multiple other[Self-healing notifier](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations) related configurations you can use to make notifier more efficient as per the use case.
```

#### Self Healing

> If Self-healing is enabled in Cruise Control and anomalies are detected, then automatic fix will start for the detected anomalies.

Reviewer 2: ```suggestion
If Self-healing is enabled in Cruise Control and anomalies are detected, then a fix for these will be attempted automatically.
```

> Cruise Control will generate an optimization proposal based on the anomaly that was detected and will try to fix it.

Reviewer 2: So is the rebalance command actually different for disk failure, broker failure, or goal failure? Does it run them with different settings?

Reviewer: There is no specific manual rebalance command for disk failure or broker failure. 
When a broker fails and cannot be recovered what you can do is actually run a rebalance with "remove-brokers" mode. If there is a disk failure and you want to move data to another disk on the same broker you could use "remove-disks" mode. In general the manual rebalance is based on the concept of goals. Depending on the goals you want to reach you can specify them and use the right mode for your use case.

> In case the anomaly is fixable, and load completeness is met then the anomaly will move to FIX_STARTED, in case anomaly is fixable then the anomaly would be IGNORED.

Reviewer 2: ```suggestion
In the case where the anomaly is fixable, and load completeness is met then the anomaly will move to FIX_STARTED, in case anomaly is fixable then the anomaly would be IGNORED.
```
How does the system know if an anomaly is fixable? Just that an OpProp has been found? 
What is "load completeness"?

> Similarly, the anomaly can move to other different states based on the factors like load completeness, goal requirement met etc.

Reviewer 2: This section needs more detail. A lot of terms have just appeared with no explanation.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
  cruiseControl:
    config:
      # configuration to enable self-healing for all anomalies
      self.healing.enabled: true  # false by default 
```

In case the user wants to enable self-healing for some specific anomalies then they can use configurations like `self.healing.broker.failure.enabled` for broker failure, `self.healing.goal.violation.enabled` for goal violation etc.

### Strimzi Notifier and Cruise Control role

`SelfHealingNotifier` is the base class which can be used when Self-healing is enabled by changing the `anomaly.notifier.class` to `com.linkedin.kafka.cruisecontrol.detector.notifier.SelfHealingNotifier`. 
The `SelfHealingNotifier` contains the logic of Self-healing and overrides multiple methods from the `AnomalyNotifier` interface giving them the implementation on what to do if certain anomalies are detected. Some of those methods are:`onGoalViolation()`
, `onBrokerFailure()`, `onDisk Failure`, `alert()` etc. The other custom notifier offered by Cruise control are also based upon the `SelfHealingNotifier`

> Since we need some way to monitor the operations that are happening in Cruise Control though the operator, we will be using `StrimziNotifier`, a new `custom` notifier

Reviewer: it's not the reason anymore we need a notifier.

To implement it, we will be extending the `SelfHealingNotifier` class and overriding the methods that we require.

> The new notifier can be easily used by adding another configuration property in the cruise control section of the Kafka custom resource called `anomaly.notifier.class`.

Reviewer 2: We will set this by default? Are we going to allow users to override this with their own notifier or are we going to lock this down (I would suggest the latter)?

Reviewer: I think we should make it forbidden and being set by the Strimzi operator to use the dedicated Strimzi notifier.

Here is an example:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
  cruiseControl:
    config:
      # configuration to enable self-healing for all anomalies
      self.healing.enabled: true  # false by default 
      anomaly.notifier.class: io.strimzi.kafka.notifier.StrimziNotifier
```

The `StrimziNotifier` will override the `alert()` method of the base `SelfHealingNotifier` class and instead of publishing the logs, it will publish Kubernetes events.

> The event published would look like this: 

Reviewer: It's the first time you show a Kube `Event` but never mentioned from where it's coming, why it's there and so on.

Reviewer 2: Yeah, you need to explain why were doing this.

```yaml
- action: DetectedAnomalyGettingFixed
  apiVersion: v1
  eventTime: "2024-12-19T08:06:21.314140Z"
  firstTimestamp: null
  involvedObject:
    kind: Pod
    name: <cruise-control-pod-name>
    namespace: myproject
  kind: Event
  lastTimestamp: null
  message: Fixing the anomaly
  metadata:
    creationTimestamp: "2024-12-19T08:06:21Z"
    generateName: cruise-control-event
    name: my-cluster-34c41626-0add-47a3-9b79-d6ee29d5d81f  # my-cluster-<Anomaly ID>
    namespace: myproject
    resourceVersion: "101070"
    uid: 80892c7d-e3a1-44ad-84dc-2d88e15a7ff0
  reason: Anomaly was detected in the cluster GOAL_VIOLATION # Detected Anomaly type
  reportingComponent: cruise-control
  reportingInstance: cruise-control
  source: {}
  type: Normal

```

The `StrimziNotifier` class is going to reside in a separate module in the operator repository called the `strimzi-notifier` so that we can later build a jar w.r.t it and then include that jar in the cruise-control image. 
This is required since we need to run the notifier as part of Cruise Control.

> We would also require to add some fabric8 client dependencies to the cruise control image since we are going to use them for creating the anomaly custom resource. <TBD - replaced by fat jar concept>

Reviewer 2: We are not creating an anomaly resource anymore? You mean the even right? Also we need to read annotations right?

### Feature Gate

> The new feature gate will be called `UseSelfHealing`.

Reviewer: You should introduce better that we want to gate the feature to allow users to test it and so on.

Reviewer 2: Agree, you need to explain why you are using the feature gate route.

It will be introduced in an alpha state and will be disabled by default.

> At this point, there is no timeline for graduation of this feature gate to beta or GA phase since it depends on things.

Reviewer 2: "Depends on things", come on Shubham you can do better than that!

Author: Sorry for this.  I have now reworked this section and will be talking with @ppatierno on how to make it better.

> The schedule will be updated later based on what the community thinks about this feature.

Reviewer 2: We could propose a schedule for graduation and let the Strimzi community comment on it in the proposal review.

### Execution Flow

1. The use should first enable the `UseSelfHealing` feature in order to use the self-healing configurations

>    * If the feature gate is disabled, self properties will be forbidden to use 

Reviewer: "self"?

> 2. After that usr should enable Self-healing in Cruise Control configuration of the Kafka custom resource by setting the `self.healing.enabled` field as `true` and also overriding the `anomaly.notifier.class` to use the new `StrimziNotifier` class. 

Reviewer 2: ```suggestion
2. After that the user should enable self-healing, in the Cruise Control configuration of the Kafka custom resource, by setting the `self.healing.enabled` field as `true` and also overriding the `anomaly.notifier.class` to use the new `StrimziNotifier` class. 
```
I would argue that user shouldn't do the second part. That should be something Strimzi does automatically.

Author: Agreed Tom

   * Once the user has enabled it, now the cluster is ready to self-heal.

>    * Cruise control will now keep a look on the cluster regarding if there is any anomaly lurking or not. 

Reviewer 2: ```suggestion

```

>    * By default, the interval at which the detector will run to detect the anomalies is 5 minutes. You can change it by using `anomaly.detection.interval.ms` property.

Reviewer 2: ```suggestion
   * By default, the interval at which the detector will run to detect the anomalies is 5 minutes. The user can change this via the `anomaly.detection.interval.ms` property.
```

3. If there is an anomaly detected, then the `StrimziNotifier` will publish a kubernetes event based on the anomaly detected.
   * The published event will contain information like anomalyID, anomalyType and anomalyProgress etc.
4. Cruise Control will then fix these detected anomalies
   * In case the anomaly is unfixable, then we will get notified about the anomaly but the fix will be ignored

### Pausing the self-healing feature

> The user will also get the flexibility to pause the self-healing feature in case they know or plan to do some rolling restart of the cluster or some rebalance they are going to perform which can cause issue with the self-healing itself.

Reviewer 2: ```suggestion
The user should also have the ability to pause the self-healing feature in case they know or plan to do some rolling restart of the cluster or some rebalance they are going to perform which can cause issue with the self-healing itself.
```
What will CC do if you try to do a rebalance when self healing is triggered?

More importantly what will happen if a user posts a `KafkaRebalance` when self-healing is running? Does the SH cancel, does the `KafkaRebalance` get set to `Ready: false` until self-healing is done?

Reviewer: Good point. Something to try. The expectation is that the executor can run just one rebalance at time, so we have to check how the endpoint replies when a new `KafkaRebalance` is created.

For this we will be having an annotation called `strimzi.io/self-healing-paused`.

> We can set this annotation on the Kafka custom resource to pause the self-healing feature. Self-healing is not paused by default so missing annotation would mean that anomalies detected will keep getting healed automatically

Reviewer 2: ```suggestion
We can set this annotation on the Kafka custom resource to pause the self-healing feature. 
Self-healing is not paused by default so missing annotation would mean that anomalies detected will keep getting healed automatically.
```

When self-healing is paused, it would mean that all the anomalies that are detected after the pause annotation is applied would be ignored and no fix would be done for them.
The anomalies that were already getting fixed would continue with the fix
Here is how you can enable this annotation:
```sh
kubectl annotate kafka my-cluster `strimzi.io/self-healing-paused: "true"`
```
The kafka custom resource will look like this:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  annotations:
     strimzi.io/self-healing-paused: "true"
  creationTimestamp: "2024-12-17T09:01:10Z"
  generation: 1
  name: my-cluster
  namespace: myproject
  resourceVersion: "109971"
  uid: 71928731-df71-4b9f-ac77-68961ac25a2f

```

> 

Reviewer 2: You haven't explained how the annotation actually prevent self-healing, its not "by magic" is it?

# Self Healing using Cruise Control

> This proposal is about integrating the Self-healing feature of Cruise Control into the Strimzi operator.

Reviewer 2: ```suggestion
This proposal is about integrating the self-healing feature of Cruise Control into the Strimzi operator.
```

Reviewer 2: "self-healing" is not a trademark, or unique item so I don't think we should treat it like a proper noun, so no need to capitalise it.

> The Self-healing feature of Cruise Control allows us to heal the anomalies detected in the cluster automatically. 

Reviewer 2: ```suggestion
The self-healing feature of Cruise Control allows us to heal anomalies, which are detected in the cluster, automatically. 
```

Anomalies are detected by Cruise Control using the anomaly detector manager.

## Current situation

During normal operations it's common for Kafka clusters to become unbalanced (through partition key skew, uneven partition placement etc.) or for problems to occur with disks or other underlying hardware which makes the cluster unhealthy.  

> Currently, if we encounter any such scenario we need to fix these issues manually i.e. if there is some broker failure then we might move the partition replicas from that corrupted broker to some other healthy broker by using the `KafkaRebalance` custom resource in the `remove-broker` mode.

Reviewer 2: Link to docs?

> While with small sized cluster it is feasible to fix things manually, when it comes to larger ones it can be very time-consuming or just not feasible to fix all the issue on your own.

Reviewer 2: ```suggestion
With smaller clusters, it is feasible to fix things manually. However, for larger ones it can be very time-consuming, or just not feasible, to fix all the issue on your own.
```

> Also, it is possible that while doing things manually, we might lose data or cause other unforeseen issues with the Kafka Cluster.

Reviewer 2: This sentence reads like you are saying "using the `KafkaRebalance` CR with Strimzi can result in data loss" which is, under normal conditions, not true. Also, given that self-healing uses the same system as "doing things manually" I am not sure if this statement would correct even if it were true that data loss was possible.

This is a long way of me saying that you should delete this sentence.

Reviewer: ... or did you see data loss in any of your tests? Which would be a big problem. Otherwise, I am with Tom about deleting this sentence.

## Motivation

> With the help of the self-healing feature we can resolve the issues like disk failure, broker failure and other issues automatically in the cluster.

Reviewer 2: ```suggestion
With the help of the self-healing feature we can resolve the issues like disk failure, broker failure and other issues automatically.
```

Cruise Control treats these issues as "anomalies", the detection of which is the responsibility of the anomaly detector mechanism.

> Currently, Strimzi has the support to enable anomaly detection by using a custom notifier. If any anomaly is detected, then that anomaly would need to be fixed manually by using the `KafkaRebalance` custom resource.

Reviewer 2: ```suggestion
Currently, Strimzi has the support to enable anomaly detection by using a custom notifier. 
If any anomaly is detected, then that anomaly would need to be fixed manually by using the `KafkaRebalance` custom resource.
```
What support does Strimzi currently provide? Is it documented anywhere?

Reviewer 2: Do you mean "currently Strimzi _could_ support the use of custom notifiers"?

Author: Yes we can currently use custom notifiers with Strimzi

Reviewer 2: ```suggestion
Currently, users are [able to set](link to docs) the notifier to one of those included with Cruise Control (<list of included notifiers>). However, any anomaly that is detected would then need to be fixed manually by using the `KafkaRebalance` custom resource.
```

It would be useful for users of Strimzi to be able to have these anomalies fixed automatically whenever they are detected.

## Proposal

### Introduction

> #### Self-healing flow diagram

Reviewer 2: This doesn't need its own section, it should just be inlined into the section below.

Author: The section below basiccally describes the information regarding different bits in the flow. I am not sure if it would be right to include this diagram into the `Anomaly Detector Manager` section.

Reviewer 2: I meant that you don't need to give it its own title. 
```suggestion
```

![Self-healing flow diagram](./images/090-self-healing-flow.png)

#### Anomaly Detector Manager

The anomaly detector manager helps in detecting the anomalies and also handling the detected anomalies.

> It acts as a coordinator between the detector classes as well as the classes which will be handling and resolving the anomalies.

Reviewer 2: Its good to have the diagram, but it doesn't show anything before the detector manager. It might be useful to add to this diagram, or have another digram, showing the sensors and detectors and how metrics trigger them.

Author: I don't see that we use metrics(sesnsors) for detecting  all the anomalies. Every detector has their unique way to detect the anomaly. If we create diagram the we will have to create multiples different flow diagrams for every anomaly

Reviewer 2: Ok, but the wider point I am making (and have made several times) is that you don't really talk anywhere about _how_ the detector classes detect anomalies. You don't need to go into great detail on each one but a short summary. For example something like: "the detector classes use various methods to detect anomalies in the Kafka cluster, including using the Cluster model from Analyser and analysing raw metrics from the metrics collector...."

> Various detector classes like `GoalViolationDetector`, `DiskFailureDetector`, `KafkaBrokerFailureDetector` etc. are used for the anomaly detection, which runs periodically to check if the cluster have their corresponding anomalies or not (This periodic check can be easily configured through the `anomaly.detection.interval.ms` configuration).

Reviewer 2: Right, but how does this anomaly detector work? Does it take derived metrics from the metrics collector and pass them to the sensor classes? Does it just pass the latest cluster model to the detector classes? 

Not saying we need a function by function description, but a bit more detail would be good I think.

Reviewer: Not all detectors work the same way. Some of them have a deep usage of requesting metadata to the Kafka cluster (i.e. broker failure), others using the admin client API (i.e. disk failure for log dirs), others uses metrics (i.e. metric anomaly) and so on. Not sure we should describe in details how each of them works, or maybe just summarizing as I stated above?

Author: @tomncooper ^^

Reviewer 2: Yes! That! Say something like that!

The detected anomalies can be of various types:

> * Goal Violation - This happens if some optimization goals are violated(for e.g. DiskUsageDistributionGoal etc.).

Reviewer 2: ```suggestion
* Goal Violation - This happens if certain optimization goals are violated (e.g. DiskUsageDistributionGoal etc.). These goals can be configured independently (through the `self.healing.goals` config) to those used for manual rebalancing.
```

* Topic Anomaly - Can happen when one or more topics in cluster violates user-defined properties (e.g. some partitions are too large in disk).

> * Broker Failure - It happens when a non-empty broker crash or leaves a cluster.

Reviewer 2: ```suggestion
* Broker Failure - It happens when a non-empty broker crashes or leaves a cluster.
```

> * Disk Failure - This failure happens if one of the non-empty disks dies(related to a Kafka Cluster with JBOD disks).

Reviewer 2: ```suggestion
* Disk Failure - This failure happens if one of the non-empty disks fails (related to a Kafka Cluster with JBOD disks).
```

> * Metric anomaly - This failure happens if metrics collected by Cruise control has some anomaly in their value(e.g. a sudden rise in log flush time metrics).

Reviewer 2: ```suggestion
* Metric anomaly - This failure happens if metrics collected by Cruise Control have some anomaly in their value (e.g. a sudden rise in the log flush time metrics).
```

> These detectors will then detect the anomalies and the sensors(metrics) with respect to those anomalies are then updated by the anomaly detector manager.

Reviewer 2: I am not sure I get what the sensors are doing here? The manager updates the sensors? I though the sensors are what the detectors read from (I haven't looked at the code in a long time so you know better than me).

Also, this is the first time you have mentioned the "sensors" and then you don't mention them again? What are they? How do they fit into the flow?

Reviewer: Maybe, but it's just my hypothesis, Shubham is referring to the "metric anomaly" detection where an anomaly is detected based just on metrics. Not sure why using "sensors" here.

Author: Yeah I was writing w.r.t `metric anomaly` as well as some of the metrics which gets updated regarding the `GoalViolation` anomalies. I will fix the sentence

> The detected anomalies are inserted in priority queue and then anomaly with the highest property is picked first to resolve.

Reviewer 2: ```suggestion
The detected anomalies are inserted into a priority queue and the anomaly with the highest property is picked first to resolve.
```

> The anomaly detector manager also calls the notifier to get an action regarding, whether the anomaly should be fixed, delayed or ignored. 

Reviewer 2: ```suggestion
The anomaly detector manager calls the notifier to get an action regarding, whether the anomaly should be fixed, delayed or ignored. 
```

> If the action is fix, then the anomaly detector manager would call the classes that are required to resolve the anomaly.

Reviewer 2: ```suggestion
If the action is fix, then the anomaly detector manager calls the classes that are required to resolve the anomaly.
```

Anomaly detection also has several other [configurations](https://github.com/linkedin/cruise-control/wiki/Configurations#anomalydetector-configurations), such as the detection interval and the anomaly notifier class, which can effect the performance of the Cruise Control server and the latency of the anomaly detection.

#### Notifiers in Cruise Control

Whenever anomalies are detected, Cruise Control provides the ability to notify the user regarding the detected anomalies using optional notifier classes.
The notification sent by these classes increases the visibility of the operation that are taken by Cruise Control.

> By default, Cruise Control uses the `NoopNotifier` which silently declines to action any detected anomaly.

Reviewer 2: I would move this line below the possible actions list and merge it with line 59.

However, the notifier class used by Cruise Control is configurable and custom notifiers can be used by setting the `anomaly.notifier.class` property.

> The notifier class also returns the `action` that is going to be taken on the notified anomaly.

Reviewer 2: ```suggestion
The notifier class returns the `action` that is going to be taken on the notified anomaly.
```

These actions can be classified into three types:
* FIX - Start the anomaly fix
* CHECK - Delay the anomaly fix
* IGNORE - Ignore the anomaly fix

The default `NoopNotifer` always sets the notifier action as `IGNORE`.

Cruise Control also provides [custom notifiers](https://github.com/linkedin/cruise-control/wiki/Configure-notifications) like Slack Notifier, Alerta Notifier etc. for notifying users regarding the anomalies. There are multiple other [Self-healing notifier](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations) related configurations you can use to make notifiers more efficient as per the use case.

#### Self Healing
If Self-healing is enabled in Cruise Control and anomalies are detected, then a fix for these will be attempted automatically.

> The action returned by the notifier would decide whether the anomaly would be fixed or not. 

Reviewer 2: ```suggestion
The action returned by the notifier would decide whether the anomaly should be fixed or not. 
```

If the notifier has returned `FIX` as the action then the classes which are responsible for resolving the anomaly would be called.

> Cruise Control will then generate an optimization proposal based on the anomaly that was detected and will try to fix it.

Reviewer 2: "based on the anomaly detected" - how does this work, are special arguments passed to CC? Does it run a full rebalance for goal violations and some other kind of rebalance for failed brokers/disks?

Reviewer: Kind of ... each detected anomaly is implemented by a specific class which leverages a corresponding other class to run a fix. For example the `BrokerFailures` anomaly class uses the `RemoveBrokersRunnable` class to fix the anomaly and so on.

Reviewer 2: So the detector class specifies the fix? Just a sentence on how the fix is applied would be nice.

> In case the anomaly detected is unfixable for e.g. violated hard goals that cannot be fixed typically due to lack of physical hardware(insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy Replica Capacity Goal, or insufficient number of resources to satisfy resource capacity goals), the anomaly wouldn't be fixed.

Reviewer 2: ```suggestion
In case the anomaly detected is unfixable, for example where violated hard goals cannot be fixed typically due to lack of physical hardware, the anomaly wouldn't be fixed.
```

Reviewer 2: What happens then? Is there a notification, a log entry, something in the UI? How would a user of raw CC be told about this?

Author: Currently we just get a warning in the cruise control logs stating that `Self-healing is not possible due to unfixable goals`

Reviewer: Unluckily we don't have a way to plug a notification here, so only existing way is the CC log. The Cruise Control community is not that open to improvements like this in my experience.

> When Self healing starts, you can check the anomaly status by polling the `state` endpoint with the `substate` configured as `anomaly_detector`. This would provide the anomaly details like anomalyId, anomalyType and the status of anomaly etc. For e.g.

Reviewer 2: ```suggestion
When self-healing starts, you can check the anomaly status by polling the `state` endpoint with the `substate` configured as `anomaly_detector`. This would provide the anomaly details like `anomalyId`, `anomalyType` and the status of anomaly etc. For example:
```

```shell

> AnomalyDetectorState: {selfHealingEnabled:[BROKER_FAILURE, DISK_FAILURE, GOAL_VIOLATION, METRIC_ANOMALY, TOPIC_ANOMALY, MAINTENANCE_EVENT], selfHealingDisabled:[], selfHealingEnabledRatio:{BROKER_FAILURE=1.0, DISK_FAILURE=1.0, GOAL_VIOLATION=1.0, METRIC_ANOMALY=1.0, TOPIC_ANOMALY=1.0, MAINTENANCE_EVENT=1.0}, recentGoalViolations:[], recentBrokerFailures:[], recentMetricAnomalies:[], recentDiskFailures:[{anomalyId=4b1ed191-50d7-4aa1-806b-b8c5c0d679fe, detectionDate=2024-12-09T09:49:46Z, statusUpdateDate=2024-12-09T09:49:47Z, failedDisksByTimeMs={2={/var/lib/kafka/data-1/kafka-log2=1733737786603}}, status=FIX_STARTED}], recentTopicAnomalies:[], recentMaintenanceEvents:[], metrics:{meanTimeBetweenAnomalies:{GOAL_VIOLATION:0.00 milliseconds, BROKER_FAILURE:0.00 milliseconds, METRIC_ANOMALY:0.00 milliseconds, DISK_FAILURE:1.31 milliseconds, TOPIC_ANOMALY:0.00 milliseconds}, meanTimeToStartFix:180.00 milliseconds, numSelfHealingStarted:1, numSelfHealingFailedToStart:0, ongoingAnomalyDuration=0.00 milliseconds}, ongoingSelfHealingAnomaly:4b1ed191-50d7-4aa1-806b-b8c5c0d679fe, balancednessScore:-1.000}

Reviewer 2: Can you run this through `jq` or just format it a bit nicer. A big blob of text isn't very readable.

```

> The anomaly status would change based on how the healing progresses. The anomaly can transition to these possible status:

Reviewer 2: ```suggestion
The anomaly status changes based on how the healing is progressing. The anomaly can transition from the initial `<initial-status>` to these possible options:
```

* DETECTED - The anomaly is just detected and no action is yet taken on it 

> * IGNORED - Self healing is either disabled/ unfixable anomaly

Reviewer 2: ```suggestion
* IGNORED - Self healing is either disabled or the anomaly is unfixable
```

* FIX_STARTED - The anomaly has started getting fixed
* FIX_FAILED_TO_START - The proposal generation for the anomaly fix failed 
* CHECK_WITH_DELAY - The anomaly fix is delayed 
* LOAD_MONITOR_NOT_READY - Monitor which is monitoring the workload of a Kafka cluster is in loading or bootstrap state

> * COMPLETENESS_NOT_READY - Completeness not met for some goals(The completeness describes the confidence level of the data in the metric sample aggregator. It is primarily measured by the validity of the metrics samples in different windows.)

Reviewer 2: ```suggestion
* COMPLETENESS_NOT_READY - Completeness not met for some goals. The completeness describes the confidence level of the data in the metric sample aggregator. If this is low then some Goal classes may refuse to produce proposals.
```

> We can also poll the `state` endpoint with `substate` as `executor` to get details on regarding the task that are being performed in the cluster. For e.g.

Reviewer 2: ```suggestion
We can also poll the `state` endpoint, with `substate` parameter as `executor`, to get details on the tasks that are currently being performed by Cruise Control. For example:
```

Reviewer 2: "e.g." is already an abbreviation of "for example" and it only tends to be used inside brackets, so I would just use the full words.

```shell

> ExecutorState: {state: INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS, pending(20)/in-progress(0)/aborting(0)/finished(0)/total(20) inter-broker partition movements, completed(0)/total(0) bytes in MBs: 0.00%, max/min/avg concurrent inter-broker partition movements per-broker: 5/5/5.00, triggeredSelfHealingTaskId: 4b1ed191-50d7-4aa1-806b-b8c5c0d679fe, triggeredTaskReason: Self healing for DISK_FAILURE: {

Reviewer 2: Again, this is a wall of text, can you format it a bit?

Author: Okay I will fix this

	Disk /var/lib/kafka/data-1/kafka-log2 on broker 2 failed at 2024-12-09T09:49:46Z
 }}
```

Self-healing is also a very robust and resilient feature due to its ability to tackle issues which can appear while the healing of cluster is happening. 

> For e.g. If we have a broker failure and self-healing is being triggered, then Cruise Control will try to move the partition replicas from the dead/offline broker to some healthy broker, and if some other broker gets deleted/rolled  between the healing, then we would be prompted through logs that `Self-healing finished successfuly with error` and then the unfixed anomaly would be detected again after the periodic anomaly check and will start to get healed again

Reviewer 2: "self healing is being triggered" do you mean the broker failure triggers the self healing or that self healing is running already and a broker fails. The sentence before this one would suggest the latter?

"some other broker gets deleted/rolled  between the healing" or do you mean "during" the healing?

This sentence is not very clear. I _think_ you mean that:

```suggestion
For example, if a broker fails, self-healing will trigger and Cruise Control will try to move the partition replicas from the failed broker to other healthy brokers in the Kafka cluster. If, during the process of moving partitions from the failed broker, another broker in the Kafka cluster gets deleted or rolled, then Cruise Control would finish the transfer from the failed broker but log that `self-healing finished successful with error`. Cruise Control would then, after the periodic anomaly check interval, detect and fix the new issue of the deleted broker.
```

Reviewer 2: Would it actually do anything with a rolled broker? If it came back up and synced within 2 anomaly detection periods then CC wouldn't care right?

In fact the issue of rolling brokers and anomaly detection probably needs its own discussion/section.

Reviewer: > If it came back up and synced within 2 anomaly detection periods then CC wouldn't care right?

If the first anomaly fix ends with errors (because CC is not able to move partitions due to a rolled/failed broker), when a broker comes back, it will be a totally new detected anomaly so CC will be able to behave accordingly (with a new fix, a new proposal and so on).

> ### Enabling Self-healing configuration in operator

Reviewer 2: ```suggestion
### Enabling Self-healing configuration in the Strimzi operator
```

> Every configuration related to Self-healing is currently disabled in the operator, and we will need to enable all those properties. The properties will be available behind a feature gate.

Reviewer 2: ```suggestion
Every configuration related to self-healing is currently disabled in the operator and we will need to enable users altering some of those properties. 
The setting of these properties will be placed behind a feature gate.
```

Reviewer 2: It probably won't be all the properties we enable. For example we might want to lock down the notifier class config.

Reviewer: +1 ... when self-healing is not enabled, we should allow the user to specify their own notifier class (as it's possible today, even because it's part of the anomaly detection not the self-healing). If the self-healing is enabled, we should lock the notifier class parameters because the CO will configure the one needed for generating events.
Unless we want to do a total freedom to the users to specify their own notifier in any case.
The previous solution leveraging a `KafkaAnomaly` custom resource needed a locked Strimzi notifier (to create such resource). The new solution is just generating Kube events and nothing more, right? So maybe the users could just plugin their notifier to do something different? It depends on how much opinionated we want to be on the provided self-healing support.

Reviewer: Of course, I forgot that our StrimziNotifier would take into account the flag to disable self-healing temporary and returning IGNORE for each anomaly :-/

Reviewer 2: If users can set their own notifiers then we will need logic to make sure that any user specified notifier setting is _combined_ with the strimzi notifier when self-healing is enabled.

### Feature Gate

> The Self-healing feature will be introduced behind a feature gate.

Reviewer 2: ```suggestion
The self-healing feature will be introduced behind a feature gate.
```

Reviewer 2: You said this already at the end of the last section, I think you can get rid of this sentence.

The new feature gate will be called `UseSelfHealing`.
It will be introduced in an alpha state and will be disabled by default.

> We are using a feature gate for this feature due to various reasons:

Reviewer 2: ```suggestion
Using a feature gate for self-healing was chosen because:
```

Reviewer 2: In my opinion, none of the reasons you list are a justification for using a feature gate instead of just adding the self-healing logic to base Strimzi. You could achieve all the reasons you list by just setting some kind of `enableSelfHealing` config? 

Why is a feature gate useful? If we don't use an FG, added the SH logic to base strimzi and used a config flag instead, would we need to change internal Strimzi logic in a potentially dangerous way that could effect other Strimzi operations even with the flag set to false? If not, then maybe a FG is overkill?

Author: Okay, I will rework this section

Reviewer: FG and a `enableSelfHealing` config are on two different levels.
With a FG you are covering all the clusters you are handling with your CO because the self-healing feature is totally disabled at ... the "core". With a `enableSelfHealing` config you are disabling it at Kafka cluster level, by leaving other clusters leveraging the self-healing instead. There are pros and cons for both about avoiding a user making mistakes on some clusters but not others, or allowing the user to try self-healing on some clusters but not others.

Reviewer 2: Well my question stands, are we changing core logic in a way that will effect vital functions even with self-healing disabled? If so then a FG is a good idea. If not, then it seems like overkill.

* We want to allow users to try this feature and test it.

> * We would get to know more about the positive/ negatives of this feature.

Reviewer 2: ```suggestion
* We would like to know more about the positive/negatives of this feature by testing it in real deployments.
```

* Getting feedback from the community would allow us to understand how we can improve this even more.

> ### Enabling Self-healing in Kafka custom resource

Reviewer 2: ```suggestion
### Enabling Self-healing in the Kafka custom resource
```

> When the feature gate is enabled, the operator would set the `self.healing.enabled` configuration to `true` and would allow the users to use all the `self-healing` related [configurations](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations).

Reviewer 2: ```suggestion
When the feature gate is enabled, the operator will set the `self.healing.enabled` Cruise Control configuration to `true` and allow the users to set other appropriate `self-healing` related [configurations](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations).
```

> Here is an example of how the configured Kafka custom resource would look like:

Reviewer 2: ```suggestion
Here is an example of how the configured Kafka custom resource could look:
```

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
  cruiseControl:
    config:
      # configuration to enable self-healing for all anomalies
      self.healing.enabled: true
```

> In case the user wants to enable Self-healing for some specific anomalies then they can use configurations like `self.healing.broker.failure.enabled` for broker failure, `self.healing.goal.violation.enabled` for goal violation etc.

Reviewer 2: ```suggestion
Where the user wants to enable self-healing for some specific anomalies, they can use configurations like `self.healing.broker.failure.enabled` for broker failure, `self.healing.goal.violation.enabled` for goal violation etc.
```

### Strimzi Notifier

> `SelfHealingNotifier` is the base class which can be used when Self-healing is enabled by changing the `anomaly.notifier.class` to `com.linkedin.kafka.cruisecontrol.detector.notifier.SelfHealingNotifier`. 

Reviewer 2: This sentence implies that self-healing is enabled by changing the notifier class? What I think you mean is:
```suggestion
Cruise Control provides a base class (`SelfHealingNotifier`) which can be used to easily implement your own notifier (set via the `anomaly.notifier.class` config). 
```

> The `SelfHealingNotifier` contains the logic of Self-healing and overrides multiple methods from the `AnomalyNotifier` interface giving them the implementation on what to do if certain anomalies are detected. Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDisk Failure`, `alert()` etc.

Reviewer 2: ```suggestion
The `SelfHealingNotifier` contains the logic of self-healing and implements multiple methods from the `AnomalyNotifier` interface providing the logic of what to do if certain anomalies are detected. 
Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDisk Failure`, `alert()` etc.
```

Reviewer 2: I think these sentences and the one above are the wrong way round, you should start by describing the interface and then introduce the concrete base class and then talk about other implementations.

Author: Okay. I will fix this

The other custom notifier offered by Cruise control are also based upon the `SelfHealingNotifier`.

> We will be using our own notifier called `StrimziNotifer` to leverage Self-healing.

Reviewer 2: ```suggestion
We will be using our own notifier implementation called `StrimziNotifer`.
```

> When Self-healing would be enabled, then Strimzi would configure Cruise Control to use the `StrimziNotifier` to notify regarding the anomalies and take action on them. 

Reviewer 2: ```suggestion
When self-healing is enabled, Strimzi will configure Cruise Control to use the `StrimziNotifier` to notify regarding the anomalies and take action on them. 
```

Reviewer 2: Is this only when self healing is enable in the CC config or would not it just make sense to have it always configured to use our notifier?

> The `StrimziNotifier` will be extending the `SelfHealingNotifier` class and overriding the methods that we require.

Reviewer 2: You don't need to say this explicitly, it is implied by the sentences above.

> Here is an example of how the configured Cruise Control config would look like:

Reviewer 2: ```suggestion
Here is an example of what the configured Cruise Control config would look like:
```

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    # ...
  cruiseControl:
    config:
      # configuration to enable self-healing for all anomalies
      self.healing.enabled: true  # false by default 

>       anomaly.notifier.class: io.strimzi.kafka.notifier.StrimziNotifier

Reviewer 2: As I have said before, the user should not need to do this. This should be invisible to them, Strimzi should set this. Unless you are proposing that users can add their own custom notifiers, this config options should not even be available.

```

> For effective alerts and monitoring, we will be making use of the `Kubernetes events`. 

Reviewer 2: ```suggestion
For effective alerts and monitoring, we will be making use of Kubernetes events to signal to users what operations Cruise Control is performing automatically. 
```

> The `StrimziNotifier` would be overriding the `alert()` method of the `SelfHealingNotifier` and will trigger and publish `events` whenever an anomaly is detected

Reviewer 2: ```suggestion
The `StrimziNotifier` will override the `alert()` method of the `SelfHealingNotifier` and will trigger and publish events whenever an anomaly is detected.
```

> These `events` would publish information regarding the anomaly like the anomalyID, the type of anomaly which was detected etc.

Reviewer 2: ```suggestion
These events would publish information regarding the anomaly like the `anomalyID`, the type of anomaly which was detected etc.
```

Why would users care about the ID? Can they use it anywhere? Is it used anywhere else, like the logs?

Author: ID's can be helpful to identify which anomalies are basically being healed by matching them with the output from the `Executor` state

Reviewer 2: Can the user see the executor state? If not then there is no point adding IDs to the events. If however they are in the logs and could be used with debugging then having them makes sense.

Author: If the user(manually) do a request to `executor` substate using the state endpoint then they can see which anomaly is being fied and that can be a good way for the user to see which anomaly is being fixed by matching it with the anomalyID

> The `events` would help the user to understand the anomalies that are present, when the anomalies appeared in their cluster and also would help them know what is going on in their clusters.

Reviewer 2: ```suggestion
The events would help the user to understand the anomalies that are present, when the anomalies appeared in their cluster and also would help them know what is going on in their clusters.
```

The event published would look like this: 

```yaml
- action: <action>
  apiVersion: v1
  eventTime: "2024-12-19T08:06:21.314140Z"
  firstTimestamp: null
  involvedObject:
    kind: Pod
    name: <cruise-control-pod-name>
    namespace: myproject
  kind: Event
  lastTimestamp: null

>   message: Fixing the anomaly

Reviewer 2: Probably need a bit more detail here.

  metadata:
    creationTimestamp: "2024-12-19T08:06:21Z"
    generateName: cruise-control-event
    name: my-cluster-34c41626-0add-47a3-9b79-d6ee29d5d81f  # my-cluster-<Anomaly ID>
    namespace: myproject
    resourceVersion: "101070"
    uid: 80892c7d-e3a1-44ad-84dc-2d88e15a7ff0
  reason: Anomaly was detected in the cluster GOAL_VIOLATION # Detected Anomaly type
  reportingComponent: cruise-control
  reportingInstance: cruise-control
  source: {}
  type: Normal
```

> The `StrimziNotifier` would also have a Kubernetes watcher which will be watching the Kafka Custom resource.

Reviewer 2: Why is this sentence and the next one here? I would remove them and talk about the watcher in the pausing sections below.

This would be required for the pause Self-healing feature which we are going to talk about in the next section.

> The `StrimziNotifier` class is going to reside in a separate module in the operator repository called the `strimzi-notifier` so that we can later build a jar w.r.t it and then include that jar in the Cruise-Control image.

Reviewer 2: ```suggestion
The `StrimziNotifier` class is going to reside in a separate module in the operator repository called the `strimzi-cruise-control-notifier` so that a seperate jar for the notifier can be built for inclusion in the Cruise Control container image.
```

Reviewer 2: Ahhhh `w.r.t` is back 😱

> This is required since we need to run the notifier as part of Cruise Control.

Reviewer 2: You have basically already said this in the sentence above.

> We would also require to create a Strimzi fat jar which will contain the fabric8 client dependencies that we are going to use for generating the Kubernetes events as well as for reading the annotation on the Kafka resource from the notifier.

Reviewer 2: ```suggestion
We will also need to include the fabric8 client dependencies in the notifier JAR that will be used for generating the Kubernetes events as well as for reading the annotation on the Kafka resource from the notifier.
```

### Pausing the Self-healing feature

> The user will also get the flexibility to pause the Self-healing feature in case they know or plan to do some rolling restart of the cluster or some rebalance they are going to perform which can cause issue with the Self-healing itself.

Reviewer 2: ```suggestion
The user should also have the option to pause the self-healing feature, in case they know or plan to do some rolling restart of the cluster or some rebalance they are going to perform which can cause issue with the Self-healing itself.
```
Hold on, this implies that there is an issue with rolling restarts and/or running a rebalance when self-healing is enabled? Is that true?

Either way, this shouldn't be the first time you mention such a thing. As I said earlier, I think you should probably have a seperate section on what will happen with posting KafkaRebalances when a self-healing fix is running and also a section on rolling during self-healing.

Reviewer: During tests I don't think that Shubham detected issues that are mentioned here, or?
Of course, the sentence is warring and seems to say the opposite.
On the other side, giving the users an advice of disabling self-healing temporary when they know that the cluster is going trough a long brokers rolling is a good thing to do imho. Just to avoid CC running several fixes which fails each time because another broker is rolling, so actually restarting the fix from scratch.

Reviewer 2: Oh I think being able to temporarily disable SH is a good idea. Its just that this sentence makes it sound like they will experience issues if they roll during SH fixes? 

As I have said before, I think you need to have a more detailed description of what happens if you post a Rebalance CR, trigger a manual roll/upgrade and other similar Strimzi operations when a SH fix is running. From our chats it sounds like: "nothing bad it will just restart later", which is good to spell out as reviewers _will_ ask. But also having that explaination gives you justification for the pausing. "We allow pausing so that CC isn't going to keep starting and failing while you roll/upgrade the cluster".

Another question you might get is: "Why, if we think this is an issue, don't we get the CO to pause SH when it does an upgrade/roll?"

Self-healing is not paused by default so missing annotation would mean that anomalies detected will keep getting healed automatically.
When Self-healing is paused, it would mean that all the anomalies that are detected after the pause annotation is applied would be ignored and no fix would be done for them.
The anomalies that were already getting fixed would continue with the fix.

#### How the pausing works?

To pause Self-healing, you will need to annotate the Kafka custom resource with the following annotation:

```sh
kubectl annotate kafka my-cluster `strimzi.io/self-healing-paused: "true"`
```
The Kafka custom resource will look like this:
```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  annotations:
     strimzi.io/self-healing-paused: "true"
  creationTimestamp: "2024-12-17T09:01:10Z"
  generation: 1
  name: my-cluster
  namespace: myproject
  resourceVersion: "109971"
  uid: 71928731-df71-4b9f-ac77-68961ac25a2f
```

Once the `annotation` is applied to the Kafka custom resource, the notifier would know about this change since we will be configuring it with a Kubernetes watcher which will be watching for any changes in the Kafka custom resource.

> We will also have checks on methods like `onGoalViolation`, `onBrokerFailure`, `onTopicAnomaly` etc. in the `StrimziNotifier` to check if `Self-healing` is paused or not.

Reviewer 2: Why would need a watcher if every time the notifier does anything it checks for the annotation anyway? Isn't it simpler to always check and then you don't need to worry about missed watcher triggers?

Reviewer: Yeah, no need for a watcher here. It would be just the notifier looking at the Kafka custom resource and check if the annotation is set or not.

If the annotation is applied then this check will set the `action` to be taken on the anomaly as `IGNORE` and the anomalies wouldn't be fixed.

### Execution Flow

1. The use should first enable the `UseSelfHealing` feature in order to use the Self-healing configurations

>    * If the feature gate is disabled, Self-healing properties will be forbidden to use

Reviewer 2: Again, why use a FG? The point of a FG is that it allows you to enable/disable code paths and whole blocks of logic. Do we need that?

Author: I didn't touch this section, since I think we will probably get rid of this section when I was having talk with @ppatierno. I will try add to add all the details in the sections above itself

> 2. After that usr should enable Self-healing in Cruise Control configuration of the Kafka custom resource by setting the `self.healing.enabled` field as `true` and also overriding the `anomaly.notifier.class` to use the new `StrimziNotifier` class.

Reviewer 2: ```suggestion
2. After that the user should enable self-healing in Cruise Control configuration of the Kafka custom resource by setting the `self.healing.enabled` field as `true` and also overriding the `anomaly.notifier.class` to use the new `StrimziNotifier` class.
```
As I said before, they shouldn't have to configure the notifier, Strimzi should do it for them. Or we should just set it as the default.

   * Once the user has enabled it, now the cluster is ready to self-heal.

>    * Cruise control will now keep a look on the cluster regarding if there is any anomaly lurking or not.

Reviewer 2: ```suggestion
```

>    * By default, the interval at which the detector will run to detect the anomalies is 5 minutes. You can change it by using `anomaly.detection.interval.ms` property.

Reviewer 2: As you haven't really explained detection at the start, it is hard to know what the implications of changing this setting are.

3. If there is an anomaly detected, then the `StrimziNotifier` will publish a kubernetes event based on the anomaly detected.
   * The published event will contain information like anomalyID, anomalyType and anomalyProgress etc.
4. Cruise Control will then fix these detected anomalies

>    * In case the anomaly is unfixable, then we will get notified about the anomaly but the fix will be ignored

Reviewer 2: Who will get notified and how? By an event? Or a condition on the Kafka CR? Seems pretty important if there is a problem that can't be fixed. How will that get surfaced to the user?
